internal/gcsx/inactive_timeout_reader.go

```diff
@@ -0,0 +1,245 @@
+type inactiveTimeoutReader struct {
+	object *gcs.MinObject
+	bucket gcs.Bucket
+
+	// The underlying GCS storage reader; nil if closed due to inactivity.
+	gcsReader gcs.StorageReader
+
+	// Total number of bytes successfully read so far.
+	seen uint64
+
+	// Requested range [start, end) from this reader.
+	reqRange gcs.ByteRange
+
+	// The read handle used for efficient reconnection for zonal bucket.
+	readHandle []byte
+
+	// Derived from the parent context, used for creating new readers and monitoring cancellation.
+	ctx    context.Context
+	cancel context.CancelFunc
+
+	// Mutex protecting internal state (mainly gcsReader & isActive).
+	mu locker.Locker
+
+	// Flag set by Read and reset by the monitor goroutine to track activity within the timeout window.
+	isActive bool
+}
+
+var (
+	ErrZeroInactivityTimeout = errors.New("ErrZeroInactivityTimeout")
+)
+
+// NewInactiveTimeoutReader creates a new gcs.StorageReader that wraps an
+// underlying GCS reader. It attempts to create the initial reader using the
+// provided parameters. If successful, it starts a background goroutine to monitor
+// for inactivity based on the specified timeout.
+//
+// If the timeout duration is zero, it returns (nil, ErrZeroInactivityTimeout) as a zero timeout
+// defeats the purpose of this wrapper.
+func NewInactiveTimeoutReader(ctx context.Context, bucket gcs.Bucket, object *gcs.MinObject, readHandle []byte, byteRange gcs.ByteRange, timeout time.Duration) (gcs.StorageReader, error) {
+	return NewInactiveTimeoutReaderWithClock(ctx, bucket, object, readHandle, byteRange, timeout, clock.RealClock{})
+}
+
+func NewInactiveTimeoutReaderWithClock(ctx context.Context, bucket gcs.Bucket, object *gcs.MinObject, readHandle []byte, byteRange gcs.ByteRange, timeout time.Duration, clock clock.Clock) (gcs.StorageReader, error) {
+	if timeout == 0 {
+		return nil, ErrZeroInactivityTimeout
+	}
+
+	itr := &inactiveTimeoutReader{
+		object:     object,
+		bucket:     bucket,
+		reqRange:   byteRange,
+		readHandle: readHandle,
+		mu:         locker.New("inactiveTimeoutReader: "+object.Name, func() {}),
+		isActive:   false,
+	}
+	itr.ctx, itr.cancel = context.WithCancel(ctx)
+
+	var err error
+	if itr.gcsReader, err = itr.createGCSReader(); err != nil {
+		return nil, err
+	}
+
+	// Start the background periodic routine.
+	go itr.monitor(clock, timeout)
+
+	return itr, nil
+}
+
+// createGCSReader is a helper method to create the underlined reader from itr.start + itr.seen offset.
+func (itr *inactiveTimeoutReader) createGCSReader() (gcs.StorageReader, error) {
+	reader, err := itr.bucket.NewReaderWithReadHandle(
+		itr.ctx,
+		&gcs.ReadObjectRequest{
+			Name:       itr.object.Name,
+			Generation: itr.object.Generation,
+			Range: &gcs.ByteRange{
+				Start: itr.reqRange.Start + itr.seen,
+				Limit: itr.reqRange.Limit,
+			},
+			ReadCompressed: itr.object.HasContentEncodingGzip(),
+			ReadHandle:     itr.readHandle,
+		})
+	if err != nil {
```

Comment:
Do we want to handle clobbered error here?
e.g.
```
var notFoundError *gcs.NotFoundError
	if errors.As(err, &notFoundError) {
		err = &gcsfuse_errors.FileClobberedError{
			Err: fmt.Errorf("NewReader: %w", err),
		}
		return
	}
```

Reply:
Nope, that will handled by the wrapper instance, that means in the random reader while integration.